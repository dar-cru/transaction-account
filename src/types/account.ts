export enum AccountStatus {
  ACTIVE = "active",
  PENDING = "pending"
}

export type Account = {
  type: string;
  id: string;
  name: string;
  status: AccountStatus;
  balance: number;
};