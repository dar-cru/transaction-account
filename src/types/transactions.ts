export enum TransactionType {
  DEBIT = "debit",
  CREDIT = "credit",
}

export type Transaction = {
  name: string;
  amount: number;
  type: TransactionType;
  dateTime: string;
};

export type TransactionResponse = {
  accountNumber: number;
  id: number;
  transactions: Transaction[];
};