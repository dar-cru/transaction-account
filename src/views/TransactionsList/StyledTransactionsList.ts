import styled from "styled-components";

export const Table = styled.table`
  width: 100%;
  border-collapse: collapse;
`;

export const TableHeading = styled.th`
  border: 1px solid #ddd;
  padding: 8px;
  padding-top: 12px;
  padding-bottom: 12px;
  text-align: left;
  background-color: #4CAF50;
  color: white;
`
export const TableRow = styled.tr`
  :nth-child(even) {
    background-color: #f2f2f2;
  }

  :hover {
    background-color: #ddd;
  }
`

export const TableData = styled.td`
  border: 1px solid #ddd;
  padding: 8px;
`
