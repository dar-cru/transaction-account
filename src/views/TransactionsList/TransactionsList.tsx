import { Transaction } from "../../types";
import {
  Table,
  TableHeading,
  TableRow,
  TableData,
} from "./StyledTransactionsList";

export type TransactionsProps = {
  transactions: Transaction[];
};

const TransactionsList = ({ transactions }: TransactionsProps) => {
  return (
    <div style={{ width: "80%", margin: "0 auto" }}>
      <Table>
        <tbody>
          <tr>
            <TableHeading>Transaction Name</TableHeading>
            <TableHeading>Amount</TableHeading>
            <TableHeading>Transaction Type</TableHeading>
            <TableHeading>Date</TableHeading>
          </tr>
          {transactions.map((transaction, i) => {
            return (
              <TableRow key={`transaction-${transaction.name}-${i}`}>
                <TableData>{transaction.name}</TableData>
                <TableData>
                  {transaction.amount.toLocaleString("en-AU", {
                    style: "currency",
                    currency: "AUD",
                  })}
                </TableData>
                <TableData>{transaction.type}</TableData>
                <TableData>{transaction.dateTime}</TableData>
              </TableRow>
            );
          })}
        </tbody>
      </Table>
    </div>
  );
};

export default TransactionsList;
