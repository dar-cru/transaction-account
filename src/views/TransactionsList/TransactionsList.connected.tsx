import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useLocation } from "react-router-dom";
import InPageAlert from "../../components/InPageAlert";
import { RootState } from "../../redux/rootReducer";
import {
  fetchTransactions,
  resetTransactions,
} from "../../slices/transactionsSlice";
import TransactionsList from "./TransactionsList";

export type TransactionListProps = {
  accountNumber: string;
};

export const Connected = () => {
  const location = useLocation();
  const accountNumber = (location?.state as TransactionListProps)
    ?.accountNumber;
  const { transactions, isLoading } = useSelector(
    (state: RootState) => state.transaction
  );
  const dispatch = useDispatch();

  useEffect(() => {
    if (accountNumber) {
      dispatch(fetchTransactions(accountNumber));
    }

    return () => {
      dispatch(resetTransactions());
    };
  }, [dispatch]);

  return (
    <>
      {isLoading ? (
        <>Loading</>
      ) : (
        <>
          {transactions?.length > 0 ? (
            <TransactionsList transactions={transactions} />
          ) : (
            <InPageAlert
              type="error"
              message="There was an error retrieving your transactions. Transactions are not available for pending accounts."
            />
          )}
        </>
      )}
    </>
  );
};

export default Connected;
