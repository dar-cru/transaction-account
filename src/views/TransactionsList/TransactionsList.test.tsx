import React from "react";
import { render, RenderResult } from "@testing-library/react";
import TransactionsList from "./TransactionsList";
import { Transaction, TransactionType } from "../../types";

const transactions: Transaction[] = [
  {
    name: "Uber Eats",
    amount: 50.22,
    type: TransactionType.DEBIT,
    dateTime: "2021-01-02",
  },
  {
    name: "Some shopping mall",
    amount: 150.22,
    type: TransactionType.DEBIT,
    dateTime: "2021-01-04",
  },
  {
    name: "Colworths Supermaket",
    amount: 350.22,
    type: TransactionType.DEBIT,
    dateTime: "2021-01-04",
  },
  {
    name: "Lotto win",
    amount: 2.5,
    type: TransactionType.CREDIT,
    dateTime: "2021-01-04",
  },
];

describe("<TransactionsList />", () => {
  test("renders the login form if there is no authentication", () => {
    const tree: RenderResult = render(<TransactionsList transactions={transactions} />);
    
    const { container } = tree;
    expect(container).toMatchSnapshot();
  });
});
