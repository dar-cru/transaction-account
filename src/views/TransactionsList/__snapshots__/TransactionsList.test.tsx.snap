// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionsList /> renders the login form if there is no authentication 1`] = `
.c0 {
  width: 100%;
}

.c1 {
  border: 1px solid #ddd;
  padding: 8px;
  padding-top: 12px;
  padding-bottom: 12px;
  text-align: left;
  background-color: #4CAF50;
  color: white;
}

.c2:nth-child(even) {
  background-color: #f2f2f2;
}

.c2:hover {
  background-color: #ddd;
}

.c3 {
  border: 1px solid #ddd;
  padding: 8px;
}

<div>
  <div
    style="width: 80%; margin: 0px auto;"
  >
    <table
      class="c0"
    >
      <tbody>
        <tr>
          <th
            class="c1"
          >
            Transaction Name
          </th>
          <th
            class="c1"
          >
            Amount
          </th>
          <th
            class="c1"
          >
            Transaction Type
          </th>
          <th
            class="c1"
          >
            Date
          </th>
        </tr>
        <tr
          class="c2"
        >
          <td
            class="c3"
          >
            Uber Eats
          </td>
          <td
            class="c3"
          >
            $50.22
          </td>
          <td
            class="c3"
          >
            debit
          </td>
          <td
            class="c3"
          >
            2021-01-02
          </td>
        </tr>
        <tr
          class="c2"
        >
          <td
            class="c3"
          >
            Some shopping mall
          </td>
          <td
            class="c3"
          >
            $150.22
          </td>
          <td
            class="c3"
          >
            debit
          </td>
          <td
            class="c3"
          >
            2021-01-04
          </td>
        </tr>
        <tr
          class="c2"
        >
          <td
            class="c3"
          >
            Colworths Supermaket
          </td>
          <td
            class="c3"
          >
            $350.22
          </td>
          <td
            class="c3"
          >
            debit
          </td>
          <td
            class="c3"
          >
            2021-01-04
          </td>
        </tr>
        <tr
          class="c2"
        >
          <td
            class="c3"
          >
            Lotto win
          </td>
          <td
            class="c3"
          >
            $2.50
          </td>
          <td
            class="c3"
          >
            credit
          </td>
          <td
            class="c3"
          >
            2021-01-04
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
